### ISSUE TITLE
[opentelemetry-cpp] Build error

### ISSUE THREAD

**Issue by m8mble** [microsoft/vcpkg#30118]
**TL;DR**

`opentelemetry-cpp` fails to build with some features. It appears to be caused by `abseil` version `20230125.0#0` not propagating the fact that `cxx17` was enabled. 

**Host Environment**

- OS: Arch Linux
- Compiler: `g++ (GCC) 12.2.1 20230201`

**To Reproduce**

```
❯ cat vcpkg.json
{
    "name": "bug-report",
    "version-string": "1.0.0",
    "dependencies": [
        {
            "name": "abseil",
            "features": [
                "cxx17"
            ]
        },
        {
            "name": "opentelemetry-cpp",
            "features": [
                "otlp",
                "otlp-http"
            ]
        }
    ]
}
❯ vcpkg install
```

**Failure logs**

<details>
  <summary>cmake output</summary>

```
❯ vcpkg install
Detecting compiler hash for triplet x64-linux...
The following packages will be built and installed:
  * grpc[codegen,core]:x64-linux -> 1.51.1
  * nlohmann-json[core]:x64-linux -> 3.11.2
    opentelemetry-cpp[core,otlp,otlp-http]:x64-linux -> 1.8.2
Additional packages (*) will be modified to complete this operation.
Restored 0 package(s) from /home/me/.cache/vcpkg/archives in 8.76 us. Use --debug to see more details.
Installing 1/3 grpc:x64-linux...
Building grpc[codegen,core]:x64-linux...
-- Using cached grpc-grpc-v1.51.1.tar.gz.
-- Cleaning sources at /home/me/software/vcpkg/buildtrees/grpc/src/v1.51.1-1066d25324.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /home/me/software/vcpkg/downloads/grpc-grpc-v1.51.1.tar.gz
-- Applying patch 00001-fix-uwp.patch
-- Applying patch 00002-static-linking-in-linux.patch
-- Applying patch 00003-undef-base64-macro.patch
-- Applying patch 00004-link-gdi32-on-windows.patch
-- Applying patch 00005-fix-uwp-error.patch
-- Applying patch 00009-use-system-upb.patch
-- Applying patch snprintf.patch
-- Applying patch 00012-fix-use-cxx17.patch
-- Applying patch 00014-pkgconfig-upbdefs.patch
-- Applying patch 00015-disable-download-archive.patch
-- Using source at /home/me/software/vcpkg/buildtrees/grpc/src/v1.51.1-1066d25324.clean
-- Found external ninja('1.11.1').
-- Configuring x64-linux
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/lib/pkgconfig/gpr.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/lib/pkgconfig/grpc++.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/lib/pkgconfig/grpc++_unsecure.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/lib/pkgconfig/grpc.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/lib/pkgconfig/grpc_unsecure.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/debug/lib/pkgconfig/gpr.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/debug/lib/pkgconfig/grpc++.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/debug/lib/pkgconfig/grpc++_unsecure.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/debug/lib/pkgconfig/grpc.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/grpc_x64-linux/debug/lib/pkgconfig/grpc_unsecure.pc
-- Installing: /home/me/software/vcpkg/packages/grpc_x64-linux/share/grpc/copyright
-- Performing post-build validation
Stored binary cache: "/home/me/.cache/vcpkg/archives/b8/b885eeae55ae21449dc725328436b78753cff59dd21d5444103ff9faf65d5e17.zip"
Elapsed time to handle grpc:x64-linux: 13.22 min
Installing 2/3 nlohmann-json:x64-linux...
Building nlohmann-json[core]:x64-linux...
-- Downloading https://github.com/nlohmann/json/archive/v3.11.2.tar.gz -> nlohmann-json-v3.11.2.tar.gz...
-- Extracting source /home/me/software/vcpkg/downloads/nlohmann-json-v3.11.2.tar.gz
-- Using source at /home/me/software/vcpkg/buildtrees/nlohmann-json/src/v3.11.2-efdad4750e.clean
-- Found external ninja('1.11.1').
-- Configuring x64-linux
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/nlohmann-json_x64-linux/share/pkgconfig/nlohmann_json.pc
-- Fixing pkgconfig file: /home/me/software/vcpkg/packages/nlohmann-json_x64-linux/debug/share/pkgconfig/nlohmann_json.pc
-- Installing: /home/me/software/vcpkg/packages/nlohmann-json_x64-linux/share/nlohmann-json/copyright
-- Performing post-build validation
Stored binary cache: "/home/me/.cache/vcpkg/archives/3b/3baeecfc705b8c32df9d2105edd4ca10b4e8e378998113ebf470418266df1d32.zip"
Elapsed time to handle nlohmann-json:x64-linux: 2.332 s
Installing 3/3 opentelemetry-cpp:x64-linux...
Building opentelemetry-cpp[core,otlp,otlp-http]:x64-linux...
-- Downloading https://github.com/open-telemetry/opentelemetry-cpp/archive/v1.8.2.tar.gz -> open-telemetry-opentelemetry-cpp-v1.8.2.tar.gz...
-- Extracting source /home/me/software/vcpkg/downloads/open-telemetry-opentelemetry-cpp-v1.8.2.tar.gz
-- Applying patch support_absl_cxx17.patch
-- Using source at /home/me/software/vcpkg/buildtrees/opentelemetry-cpp/src/v1.8.2-19a2a784eb.clean
-- Downloading https://github.com/open-telemetry/opentelemetry-proto/archive/v0.19.0.tar.gz -> opentelemetry-proto-0.19.0.tar.gz...
-- Extracting source /home/me/software/vcpkg/downloads/opentelemetry-proto-0.19.0.tar.gz
-- Using source at /home/me/software/vcpkg/buildtrees/opentelemetry-cpp/src/ry-proto-0-df22e1b46a.clean
-- Found external ninja('1.11.1').
-- Configuring x64-linux
-- Building x64-linux-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: /usr/bin/cmake --build . --config Debug --target install -- -v -j17
    Working Directory: /home/me/software/vcpkg/buildtrees/opentelemetry-cpp/x64-linux-dbg
    See logs for more information:
      /home/me/software/vcpkg/buildtrees/opentelemetry-cpp/install-x64-linux-dbg-out.log

Call Stack (most recent call first):
  /tmp/vcpkg-bug/vcpkg_installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  /tmp/vcpkg-bug/vcpkg_installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/opentelemetry-cpp/portfile.cmake:52 (vcpkg_cmake_install)
  scripts/ports.cmake:147 (include)


error: building opentelemetry-cpp:x64-linux failed with: BUILD_FAILED
error: Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
    https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+opentelemetry-cpp
You can submit a new issue at:
    https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[opentelemetry-cpp]+Build+error
Include '[opentelemetry-cpp] Build error' in your bug report title, the following version information in your bug description, and attach any relevant failure logs from above.
    vcpkg-tool version: 2022-12-14-7ae0d8527fb488fde10a89c2813802dc9b03b6f9
    vcpkg-scripts version: dd1b57240 2023-03-09 (3 hours ago)


Please use the prefilled template from /tmp/vcpkg-bug/vcpkg_installed/vcpkg/issue_body.md when reporting your issue.
```
</details>
<details>
<summary><code>/home/me/software/vcpkg/buildtrees/opentelemetry-cpp/config-x64-linux-err.log</code></summary>
empty
</details>

[`/home/me/software/vcpkg/buildtrees/opentelemetry-cpp/config-x64-linux-out.log`](https://github.com/microsoft/vcpkg/files/10942300/config-x64-linux-out.log)
[`/home/me/software/vcpkg/buildtrees/opentelemetry-cpp/install-x64-linux-dbg-out.log`](https://github.com/microsoft/vcpkg/files/10942302/install-x64-linux-dbg-out.log)

**Additional context**

I debuged the issue a little. It is a specific issue with the latest abseil port. If I downgrade to `20220623.1#0`, the issue doesn't appear (i.e. `vcpkg install` finishes successfully).

Notice that the config log doesn't show [`Found absl uses CXX17, enable CXX17 feature.`](https://github.com/microsoft/vcpkg/blob/master/ports/opentelemetry-cpp/support_absl_cxx17.patch#L10), i.e. `ABSL_USE_CXX17` isn't set for `opentelemetry-cpp`. That leads me to believe, this is actually a bug in the latest abseil port.

**Comment by jimwang118** [comment]
Host Environment

> Ubuntu 19.04
> gcc version 8.3.0 

To Reproduce

Steps to reproduce the behavior:
`./vcpkg install opentelemetry-cpp`

Result：
![image](https://user-images.githubusercontent.com/122244446/224639167-54316a1d-6922-4fa3-9219-2f5a54b18b7a.png)
I'm not sure if the results are different due to differences between my installation environment and yours.

**Comment by m8mble** [comment]
Installing `opentelemetry-cpp` works fine for me too. This builds `abseil` with default features. But my issue appears when installing `abseil[cxx17]` with `opentelemetry-cpp`. 

I don't know how to trigger this in classic mode. My report includes a reproduction config to trigger the reported behavior in manifest mode.

**Comment by jimwang118** [comment]
There is no `abseil `in the feature of opentelemetry-cpp, why do you add `abseil `in the feature ?
You can take a look at this PR https://github.com/microsoft/vcpkg/pull/29974 which has fixed this issue.

**Comment by m8mble** [comment]
I'm building an application that depends on ports `abseil[cxx17]` as well as `opentelemetry-cpp[otlp,otlp-http]`. Therefore I'm using a manifest similar to
```
{
    "name": "bug-report",
    "version-string": "1.0.0",
    "dependencies": [
        {
            "name": "abseil",
            "features": [
                "cxx17"
            ]
        },
        {
            "name": "opentelemetry-cpp",
            "features": [
                "otlp",
                "otlp-http"
            ]
        }
    ]
}
```
Note the `opentelemetry-cpp` depends on `abseil` transitively (i.e. `opentelemetry-cpp[otlp]` depends on `grpc`, which depends on `abseil`). 

How does the reference PR forward `ABSL_USE_CXX17`?

**Comment by jimwang118** [comment]
If the C++17 standard must be used, it is recommended to add the compilation option set (CMAKE_CXX_STANDARD 17) in cmakelist.txt and complete it by patch. It is forbidden to add C++ compilation standard in feature. You can refer to this PR https://github.com/microsoft/vcpkg/pull/30466

**Comment by 0xfeedface** [comment]
The `absl` port unfortunately does CXX17 selection via a feature.

The problem seems to be that the `ABSL_USE_CXX17` variable is only set when `absl` is built in the same (CMake) configuration as `opentelemetry-cpp`. If Abseil comes from e.g. a binary cache or is already installed the `opentelemetry-cpp` build won't pick up the fact that `absl` was built in CXX17 mode.

I have a patch that modifies the `openetelemetry-cpp` build to check whether `absl` was built in CXX17 mode (i.e. uses `std::string_view`) and if so, enables CXX17 also for the `opentelemetry-cpp`. Would you accept this as a fix?

**Comment by klalumiere** [comment]
I had the same error and I confirm that #30466 solved the problem for me.

**Comment by jimwang118** [comment]
@0xfeedface, can you solve your current problem according to the method of PR https://github.com/microsoft/vcpkg/pull/30466? If it can be resolved then we can close this issue.

**Comment by smartnet-club** [comment]
@jimwang118 PR #30466 does not resolve the issue. PR #30720 looks good. Please accept it asap.

**Comment by jimwang118** [comment]
@m8mble Did the PR https://github.com/microsoft/vcpkg/pull/30720 solve your problem?

**Comment by CoderPoet** [comment]
```
{
  "name": "bug-report",
  "version": "0.1.0",
  "dependencies": [
    {
      "name": "abseil",
      "features": [
        "cxx17"
      ]
    },
    {
      "name": "opentelemetry-cpp",
      "features": [
        "otlp",
        "otlp-http",
        "zpages",
        "prometheus"
      ]
    }
  ]
}
```

I'm also the same vcpkg config, but I seem to be having other issues：

``` 
/tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/absl/meta/type_traits.h:67:8: note: candidates are: ‘template<class T> struct absl::lts_20230125::is_trivially_move_assignable’
   67 | struct is_trivially_move_assignable;
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/base/internal/invoke.h:42,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/utility/utility.h:50,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/types/variant.h:46,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/variant.h:52,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/common/attribute_value.h:10,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/observer_result.h:6,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/async_instruments.h:6,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/noop.h:6,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/provider.h:10,
                 from /tmp/new/microscope-agent/include/agent/collection/tcp_tracer/tcp_tracer_metrics.h:5,
                 from /tmp/new/microscope-agent/include/agent/collection/tcp_tracer/tcp_tracer_collection.h:12,
                 from /tmp/new/microscope-agent/src/bootstrap/bootstrap.cc:8:
/tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/meta/type_traits.h:59:8: note:                 ‘template<class T> struct absl::otel_v1::is_trivially_move_assignable’
   59 | struct is_trivially_move_assignable;
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/meta/type_traits.h:98:41: error: expected template-name before ‘<’ token
   98 |     : absl::is_trivially_move_assignable<T>::type {};
      |                                         ^
/tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/meta/type_traits.h:98:41: error: expected ‘{’ before ‘<’ token
/tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/meta/type_traits.h:102:13: error: reference to ‘is_trivially_move_assignable’ is ambiguous
  102 |     : absl::is_trivially_move_assignable<T>::type {};
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/absl/base/casts.h:38,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/absl/flags/internal/flag.h:32,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/absl/flags/flag.h:39,
                 from /tmp/new/microscope-agent/include/agent/bootstrap/bootstrap.h:7,
                 from /tmp/new/microscope-agent/src/bootstrap/bootstrap.cc:5:
/tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/absl/meta/type_traits.h:67:8: note: candidates are: ‘template<class T> struct absl::lts_20230125::is_trivially_move_assignable’
   67 | struct is_trivially_move_assignable;
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/base/internal/invoke.h:42,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/utility/utility.h:50,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/types/variant.h:46,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/variant.h:52,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/common/attribute_value.h:10,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/observer_result.h:6,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/async_instruments.h:6,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/noop.h:6,
                 from /tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/metrics/provider.h:10,
                 from /tmp/new/microscope-agent/include/agent/collection/tcp_tracer/tcp_tracer_metrics.h:5,
                 from /tmp/new/microscope-agent/include/agent/collection/tcp_tracer/tcp_tracer_collection.h:12,
                 from /tmp/new/microscope-agent/src/bootstrap/bootstrap.cc:8:
/tmp/new/microscope-agent/cmake-build-debug/vcpkg_installed/x64-linux/include/opentelemetry/nostd/internal/absl/meta/type_traits.h:59:8: note:                 ‘template<class T> struct absl::otel_v1::is_trivially_move_assignable’
   59 | struct is_trivially_move_assignable;
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
